modules/likes/queuehandler.js

```diff
@@ -213,66 +223,155 @@ jQuery( document ).click( function( e ) {
 });
 
 function JetpackLikesWidgetQueueHandler() {
-	var $wrapper, wrapperID, found;
+	var wrapperID;
+
 	if ( ! jetpackLikesMasterReady ) {
 		setTimeout( JetpackLikesWidgetQueueHandler, 500 );
 		return;
 	}
 
-	if ( jetpackLikesWidgetQueue.length > 0 ) {
-		// We may have a widget that needs creating now
-		found = false;
-		while( jetpackLikesWidgetQueue.length > 0 ) {
-			// Grab the first member of the queue that isn't already loading.
-			wrapperID = jetpackLikesWidgetQueue.splice( 0, 1 )[0];
-			if ( jQuery( '#' + wrapperID ).hasClass( 'jetpack-likes-widget-unloaded' ) ) {
-				found = true;
-				break;
-			}
-		}
-		if ( ! found ) {
-			setTimeout( JetpackLikesWidgetQueueHandler, 500 );
-			return;
-		}
-	} else if ( jQuery( 'div.jetpack-likes-widget-unloaded' ).length > 0 ) {
+	// Restore widgets to initial unloaded state when they are scrolled out of view.
+	jetpackUnloadScrolledOutWidgets();
+
+	var unloadedWidgetsInView = jetpackGetUnloadedWidgetsInView();
+
+	if ( unloadedWidgetsInView.length > 0 ) {
 		// Grab any unloaded widgets for a batch request
 		JetpackLikesBatchHandler();
+	}
 
-		// Get the next unloaded widget
-		wrapperID = jQuery( 'div.jetpack-likes-widget-unloaded' ).first()[0].id;
-		if ( ! wrapperID ) {
-			// Everything is currently loaded
-			setTimeout( JetpackLikesWidgetQueueHandler, 500 );
-			return;
+	for ( var i=0; i <= unloadedWidgetsInView.length - 1; i++ ) {
```

Comment:
it's a basic trick, but pretty standard since JS has to compute `length` here on every iteration…

```js
for ( var i = 0, length = unloadedWidgetsInView; i <= length - 1; i++ ) {
	…
}
```